src/QirTools/pyqir/qirlib/src/emit/mod.rs

```diff
@@ -0,0 +1,276 @@
+use inkwell::values::BasicValueEnum;
+
+use super::interop::SemanticModel;
+use std::collections::BTreeMap;
+use std::path::Path;
+
+pub mod constants;
+    mod intrinsics;
+pub mod qir;
+mod runtime_library;
+pub mod types;
+
+pub struct Emitter {}
+impl Emitter {
+    pub fn write(model: &SemanticModel, file_name: &str) -> Result<(), String> {
+        let ctx = inkwell::context::Context::create();
+        let context = Context::new(&ctx, model.name.as_str());
+
+        Emitter::build_entry_function(&context, model)?;
+        context.emit_ir(file_name);
+
+        Ok(())
+    }
+
+    pub fn get_ir_string(model: &SemanticModel) -> Result<String, String> {
+        let ctx = inkwell::context::Context::create();
+        let context = Context::new(&ctx, model.name.as_str());
+
+        Emitter::build_entry_function(&context, model)?;
+
+        Ok(context.get_ir_string())
+    }
+
+    pub fn build_entry_function(context: &Context<'_>, model: &SemanticModel) -> Result<(), String> {
+        let entrypoint = qir::get_entry_function(context);
+
+        let entry = context.context.append_basic_block(entrypoint, "entry");
+        context.builder.position_at_end(entry);
+
+        let qubits = Emitter::write_qubits(&model, context);
+
+        let registers = Emitter::write_registers(&model, context);
+
+        Emitter::write_instructions(&model, context, &qubits, &registers);
+
+        Emitter::free_qubits(context, &qubits);
+
+        let output = registers.get("results").unwrap();
+        context.builder.build_return(Some(&output.0));
+
+        if let Err(err) = context.module.verify() {
+            let message = err.to_string();
+            return Err(message);
+        }
+        Ok(())
+    }
+
+    fn free_qubits<'ctx>(context: &Context<'ctx>, qubits: &BTreeMap<String, BasicValueEnum<'ctx>>) {
+        for (_, value) in qubits.iter() {
+            qir::qubits::emit_release(context, value);
+        }
+    }
+
+    fn write_qubits<'ctx>(
+        model: &SemanticModel,
+        context: &Context<'ctx>,
+    ) -> BTreeMap<String, BasicValueEnum<'ctx>> {
+        let mut qubits = BTreeMap::new();
+        for reg in model.qubits.iter() {
+            let indexed_name = format!("{}{}", &reg.name[..], reg.index);
+            let value = qir::qubits::emit_allocate(&context, indexed_name.as_str());
+            qubits.insert(indexed_name, value);
+        }
+        qubits
+    }
+
+    fn write_registers<'ctx>(
+        model: &SemanticModel,
+        context: &Context<'ctx>,
+    ) -> BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)> {
+        let mut registers = BTreeMap::new();
+        let number_of_registers = model.registers.len() as u64;
+        if number_of_registers > 0 {
+            let results =
+                qir::array1d::emit_array_allocate1d(&context, 8, number_of_registers, "results");
+            registers.insert(String::from("results"), (results, None));
+            let mut sub_results = vec![];
```

Comment:
Is it possible to avoid mutability here by using `.map(|reg| ...).collect()` on `model.registers.iter`?

Reply:
```rust
        let qubits = model
            .qubits
            .iter()
            .map(|reg| {
                let indexed_name = format!("{}{}", &reg.name[..], reg.index);
                let value = qir::qubits::emit_allocate(&context, indexed_name.as_str());
                (indexed_name, value)
            })
            .collect();

        qubits
```